File: csharp/MarsRover/MarsRover.Domain/Models/MarsRover.cs
using Space.Helpers;

namespace Space.Models;

public class MarsRover
{
    public Direction Heading { get; private set; } = Direction.North;
    public Position Position { get; private set; }
    public MoveStatus Status { get; private set; }

    public MarsRover(int x, int y)
    {
        Position = new Position(x, y);
    }

    public string Report()
    {
        return $"{Position.X} {Position.Y} {Heading}";
    }

    public void TurnRight()
    {
        Heading = DirectionExtension.RotateRight(Heading);
    }

    public void TurnLeft()
    {
        Heading = DirectionExtension.RotateLeft(Heading);
    }

    public void MoveForward(Position position)
    {
        Position = Position with { X = position.X, Y = position.Y };
    }
}

File: typescript/mars-rover-ts/src/models/marsrover.ts
import { Direction, rotate } from "../helpers/direction.js";
import { directionalDelta } from "../helpers/movementExtension.js";
import { Position } from "./position.js";

export type RoverLocation = {
  coordinates: Position;
  heading: Direction;
};

export class MarsRover {
  heading: Direction;
  position: Position;

  constructor(position: Position) {
    this.position = position;
    this.heading = Direction.North;
  }

  report(): RoverLocation {
    return {
      coordinates: { x: this.position.x, y: this.position.y },
      heading: this.heading,
    };
  }

  turnRight(): void {
    this.heading = rotate(this.heading, "R");
  }

  turnLeft(): void {
    this.heading = rotate(this.heading, "L");
  }

  advanceTo(next: Position): void {
    this.position = next;
  }
}
